### PR TITLE
[Dynamic Buffer Calc] Support dynamic buffer calculation on top of port auto negotiation

### PR DESCRIPTION
**Pull request by stephenxs**
**What I did**

Update the speed used for headroom calculating according to port auto-negotiation change.
The speed taken into the calculation is now called `effective speed` in the code.

Signed-off-by: Stephen Sun <stephens@nvidia.com>

**Why I did it**

**How I verified it**
Run regression test and vs test.

**Details if related**
One logic was to take the configured speed into account when calculating the headroom because it's the maximum possible speed at which a port runs.
As port auto-negotiation supported, this logic needs to be updated accordingly:
The maximum possible speed is no longer the configured one with auto-negotiation enabled.
So the speed taken into account should be:
- If port auto-negotiation is enabled:
  maximum speed in adv_speeds if it is configured
  maximum speed in supported_speeds otherwise
- Otherwise: the configured speed

### CODE REVIEW

**Comment by stephenxs** [comment]
> @stephenxs , please add vs tests

@neethajohn i didn’t add vs test cases because 
- the supported_speeds, which is a must for port auto neg, isn’t supported by vs SAI. This means we can not run vs test for this feature.
- Other flows are covered by the existing vs test cases.

we do have a regression test for it.

**Comment by stephenxs** [comment]
Regression test: https://github.com/Azure/sonic-mgmt/pull/3647

**Comment by stephenxs** [comment]
vs failed by dynamic port break down and vlan.
```
2021-06-10T08:54:29.8721498Z test_dpb_arp_flush_on_port_oper_shut failed (1 runs remaining out of 2).
2021-06-10T08:54:29.8721951Z 	<class 'AssertionError'>
2021-06-10T08:54:29.8722684Z 	Unexpected number of keys: expected=3, received=2 (('oid:0x60000000005a7', 'oid:0x60000000010ef')), table="ASIC_STATE:SAI_OBJECT_TYPE_ROUTER_INTERFACE"
2021-06-10T08:54:29.8723493Z 	[<TracebackEntry /home/vsts/work/1/s/tests/test_port_dpb_system.py:761>, <TracebackEntry /home/vsts/work/1/s/tests/test_port_dpb_system.py:39>, <TracebackEntry /home/vsts/work/1/s/tests/dvslib/dvs_database.py:366>]
2021-06-10T08:54:29.8724124Z test_dpb_arp_flush_on_port_oper_shut failed; it passed 0 out of the required 1 times.
2021-06-10T08:54:29.8724592Z 	<class 'AssertionError'>
2021-06-10T08:54:29.8725314Z 	Unexpected number of keys: expected=3, received=2 (('oid:0x60000000005a7', 'oid:0x60000000010ef')), table="ASIC_STATE:SAI_OBJECT_TYPE_ROUTER_INTERFACE"
2021-06-10T08:54:29.8726117Z 	[<TracebackEntry /home/vsts/work/1/s/tests/test_port_dpb_system.py:761>, <TracebackEntry /home/vsts/work/1/s/tests/test_port_dpb_system.py:39>, <TracebackEntry /home/vsts/work/1/s/tests/dvslib/dvs_database.py:366>]
2021-06-10T08:54:29.8726840Z test_dpb_arp_flush_on_vlan_member_remove failed (1 runs remaining out of 2).
2021-06-10T08:54:29.8727288Z 	<class 'AssertionError'>
2021-06-10T08:54:29.8728003Z 	Unexpected number of keys: expected=3, received=2 (('oid:0x60000000005a7', 'oid:0x60000000010ef')), table="ASIC_STATE:SAI_OBJECT_TYPE_ROUTER_INTERFACE"
2021-06-10T08:54:29.8728806Z 	[<TracebackEntry /home/vsts/work/1/s/tests/test_port_dpb_system.py:825>, <TracebackEntry /home/vsts/work/1/s/tests/test_port_dpb_system.py:39>, <TracebackEntry /home/vsts/work/1/s/tests/dvslib/dvs_database.py:366>]
```
and
```
2021-06-10T08:54:29.8763084Z test_AddPortChannelToVlan failed (1 runs remaining out of 2).
2021-06-10T08:54:29.8763498Z 	<class 'AssertionError'>
2021-06-10T08:54:29.8763752Z 	assert 2 == 4
2021-06-10T08:54:29.8764144Z   -2
2021-06-10T08:54:29.8764333Z   +4
2021-06-10T08:54:29.8764626Z 	[<TracebackEntry /home/vsts/work/1/s/tests/test_vlan.py:230>]
2021-06-10T08:54:29.8764998Z test_AddPortChannelToVlan passed 1 out of the required 1 times. Success!
```

File: cfgmgr/buffermgrdyn.cpp [hunk 1]
```diff
@@ -268,11 +270,82 @@ string BufferMgrDynamic::getDynamicProfileName(const string &speed, const string
...
+    auto &&speedVec = tokenize(speedList, ',');
+    unsigned long maxSpeedNum = 0, speedNum;
+    string maxSpeedStr;
+    for (auto &speedStr : speedVec)
+    {
+        speedNum = atol(speedStr.c_str());
+        if (speedNum > maxSpeedNum)
+        {
+            maxSpeedNum = speedNum;
+            maxSpeedStr = speedStr;
```
Comment (neethajohn) [inline cfgmgr/buffermgrdyn.cpp:284]: would be better to get rid of this var and convert to string just before returning

Comment (stephenxs) [reply cfgmgr/buffermgrdyn.cpp]: Updated.

**Comment by stephenxs** [comment]
This PR depends on Azure/sonic-swss#1781 to be merged first.

**Comment by stephenxs** [comment]
vs test failed due to time out. Looks like an environmental issue. Building kernel ran into interactive configuration phase.
Need retry.

**Review by neethajohn — DISMISSED** [review]

**Comment by stephenxs** [comment]
Fix conflicts by merging master back.

**Comment by stephenxs** [comment]
Looks like the LGTM C++ got stuck. It has been running for half a day.
Need to retrigger it.

**Review by neethajohn — APPROVED** [review]

**Pull request merged**